File: LeetCodeProblemsGo/LeetCode2141MaximumRunningTimeOfNComputers/LeetCode2141MaximumRunningTimeOfNComputers.go
package LeetCode2141MaximumRunningTimeOfNComputers

func maxRunTime(n int, batteries []int) int64 {
	var totalPower int64
	for _, battery := range batteries {
		totalPower += int64(battery)
	}
	var minTarget int64 = 1
	maxTarget := totalPower / int64(n)
	var midTarget int64
	var target int64 = 0
	var availablePower int64
	for minTarget <= maxTarget {
		midTarget = (maxTarget + minTarget) >> 1
		availablePower = 0
		for _, battery := range batteries {
			availablePower += min(midTarget, int64(battery))
		}
		if availablePower >= int64(n)*midTarget {
			target = max(target, midTarget)
			minTarget = midTarget + 1
			continue
		}
		maxTarget = midTarget - 1
	}
	return target

}

File: LeetCodeProblemsCMake/Problems/LeetCode2141MaximumRunningTimeOfNComputers.cpp
//
// Created by saeed on 12/1/25.
//

#include "LeetCode2141MaximumRunningTimeOfNComputers.h"

#include <algorithm>
#include <iostream>
#include <numeric>
using namespace std;

long long LeetCode2141MaximumRunningTimeOfNComputers::maxRunTime(int n, vector<int> &batteries) {
    long long totalPower = accumulate(batteries.begin(), batteries.end(), 0LL);
    long long maxTarget = totalPower / n;
    long long minTarget = 1;
    long long target{};
    long long midTarget{};
    long long power{};
    long long desiredPower{};
    while (minTarget <= maxTarget) {
        midTarget = (minTarget + maxTarget) >> 1;
        power = 0;
        for (auto battery: batteries) {
            power += min(static_cast<long long> (battery), midTarget);
        }
        desiredPower = n;
        desiredPower *= midTarget;
        if (power >= desiredPower) {
            target = max(midTarget, target);
            minTarget = midTarget + 1;
            continue;
        }
        maxTarget = midTarget - 1;
    }
    return target;
}

void LeetCode2141MaximumRunningTimeOfNComputers::Run() {
    std::vector<int> batteries = {5, 5, 4};
    int n = 2;
    cout << maxRunTime(n, batteries) << '\n';
}
